plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp

```diff
@@ -2011,6 +2032,48 @@ void CudaCalcAmoebaMultipoleForceKernel::getInducedDipoles(ContextImpl& context,
     }
 }
 
+
+void CudaCalcAmoebaMultipoleForceKernel::getTotalDipoles(ContextImpl& context, vector<Vec3>& dipoles) {
+    ensureMultipolesValid(context);
+    int numParticles = cu.getNumAtoms();
+    dipoles.resize(numParticles);
+    const vector<int>& order = cu.getAtomIndex();
+    if (cu.getUseDoublePrecision()) {
+        vector<double4> posqVec;
+        vector<double> labDipoleVec;
+        vector<double> inducedDipoleVec;
+        double totalDipoleVecX;
+        double totalDipoleVecY;
+        double totalDipoleVecZ;
+        inducedDipole->download(inducedDipoleVec);
+        labFrameDipoles->download(labDipoleVec);
+        cu.getPosq().download(posqVec);
+        for (int i = 0; i < numParticles; i++) {
+            totalDipoleVecX = posqVec[i].x * posqVec[i].w + labDipoleVec[3*i] + inducedDipoleVec[3*i];
+            totalDipoleVecY = posqVec[i].y * posqVec[i].w + labDipoleVec[3*i+1] + inducedDipoleVec[3*i+1];
+            totalDipoleVecZ = posqVec[i].z * posqVec[i].w + labDipoleVec[3*i+2] + inducedDipoleVec[3*i+2];
```

Comment:
​The reason I included the x*q term, is that when I write out the per atom
dipole moments for each atom, and then sum over all the atoms, I should be
able to reproduce the total dipole moment of the molecule. Ideally I
understand, that should be (x-xcm)​, but for a net neutral molecule, which
I was working with, the xcm contribution was zero. I could change this
function to just have the lab frame permanent dipole and induced dipole,
and add the monopole contribution manually in my python script on top, if
you feel that that would be a better way to do it.

Thanks.

Saurabh

On 5 July 2016 at 11:23, peastman <notifications@github.com> wrote:

> In plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp
> <https://github.com/pandegroup/openmm/pull/1534#discussion_r69612932>:
>
> > +    dipoles.resize(numParticles);
> > +    const vector<int>& order = cu.getAtomIndex();
> > +    if (cu.getUseDoublePrecision()) {
> > +        vector<double4> posqVec;
> > +        vector<double> labDipoleVec;
> > +        vector<double> inducedDipoleVec;
> > +        double totalDipoleVecX;
> > +        double totalDipoleVecY;
> > +        double totalDipoleVecZ;
> > +        inducedDipole->download(inducedDipoleVec);
> > +        labFrameDipoles->download(labDipoleVec);
> > +        cu.getPosq().download(posqVec);
> > +        for (int i = 0; i < numParticles; i++) {
> > +            totalDipoleVecX = posqVec[i].x * posqVec[i].w + labDipoleVec[3*i] + inducedDipoleVec[3*i];
> > +            totalDipoleVecY = posqVec[i].y * posqVec[i].w + labDipoleVec[3*i+1] + inducedDipoleVec[3*i+1];
> > +            totalDipoleVecZ = posqVec[i].z * posqVec[i].w + labDipoleVec[3*i+2] + inducedDipoleVec[3*i+2];
>
> Is that really what you want to be computing? You include the lab frame
> fixed dipole moment, and the induced dipole moment. But then you also add
> x*q, which is the dipole moment of a point charge relative to the origin.
> That isn't what I would have expected it to return. I would expect
> getTotalDipoles() to simply return the sum of the values from
> getInducedDipoles() and getLabFramePermanentDipoles(). That is, I would
> expect the dipole moment of each particle relative to its own position, not
> relative to the origin.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandegroup/openmm/pull/1534/files/a62b6bff8ad247678dd74eadfb749c63bb896bcb#r69612932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ANPF-_chdPDQTdZ2Ql6eeqBo8_M-J85Rks5qSqETgaJpZM4JCpXy>
> .
>